runtime/src/main/java/com/datastax/oss/quarkus/runtime/metrics/MicroProfileMetricsFactory.java

```diff
@@ -0,0 +1,122 @@
+ */
+package com.datastax.oss.quarkus.runtime.metrics;
+
+import com.datastax.dse.driver.api.core.metrics.DseNodeMetric;
+import com.datastax.dse.driver.api.core.metrics.DseSessionMetric;
+import com.datastax.oss.driver.api.core.config.DefaultDriverOption;
+import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
+import com.datastax.oss.driver.api.core.metadata.Node;
+import com.datastax.oss.driver.api.core.metrics.DefaultNodeMetric;
+import com.datastax.oss.driver.api.core.metrics.DefaultSessionMetric;
+import com.datastax.oss.driver.api.core.metrics.Metrics;
+import com.datastax.oss.driver.api.core.metrics.NodeMetric;
+import com.datastax.oss.driver.api.core.metrics.SessionMetric;
+import com.datastax.oss.driver.internal.core.context.InternalDriverContext;
+import com.datastax.oss.driver.internal.core.metrics.MetricsFactory;
+import com.datastax.oss.driver.internal.core.metrics.NodeMetricUpdater;
+import com.datastax.oss.driver.internal.core.metrics.NoopNodeMetricUpdater;
+import com.datastax.oss.driver.internal.core.metrics.NoopSessionMetricUpdater;
+import com.datastax.oss.driver.internal.core.metrics.SessionMetricUpdater;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.apache.commons.lang3.NotImplementedException;
+import org.eclipse.microprofile.metrics.MetricRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MicroProfileMetricsFactory implements MetricsFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(MicroProfileMetricsFactory.class);
+
+  private final String logPrefix;
+  private final InternalDriverContext context;
+  private final Set<NodeMetric> enabledNodeMetrics;
+  private final MetricRegistry registry;
+  private final SessionMetricUpdater sessionUpdater;
+
+  public MicroProfileMetricsFactory(InternalDriverContext context, MetricRegistry metricRegistry) {
+    this.logPrefix = context.getSessionName();
+    this.context = context;
+
+    DriverExecutionProfile config = context.getConfig().getDefaultProfile();
+    Set<SessionMetric> enabledSessionMetrics =
+        parseSessionMetricPaths(config.getStringList(DefaultDriverOption.METRICS_SESSION_ENABLED));
+    this.enabledNodeMetrics =
+        parseNodeMetricPaths(config.getStringList(DefaultDriverOption.METRICS_NODE_ENABLED));
+
+    if (enabledSessionMetrics.isEmpty() && enabledNodeMetrics.isEmpty()) {
+      LOG.debug("[{}] All metrics are disabled.", logPrefix);
+      this.registry = null;
+      this.sessionUpdater = NoopSessionMetricUpdater.INSTANCE;
+    } else {
+      this.registry = metricRegistry;
+      this.sessionUpdater =
+          new MicroProfileSessionMetricUpdater(enabledSessionMetrics, registry, context);
+    }
+  }
+
+  @Override
+  public Optional<Metrics> getMetrics() {
+    throw new NotImplementedException("getMetrics() is not supported by the MicroProfile");
```

Comment:
My first instinct was to make this return `Optional.empty()`, but I think you're right, an exception is more precise.

Is using commons-lang's `NotImplementedException` a common thing in Quarkus? I think `UnsupportedOperationException` would be clear enough.

Also maybe tell users where to look if they want to access the metrics programmatically:
```
getMetrics() is not supported in Quarkus. The driver publishes its metrics directly to the vendor MetricRegistry of the container
```

Reply:
The empty option type is returned in cases where metrics are disabled, though... is this case really _that_ different?  I could see a pretty credible argument for a log line containing the info above + a return of an empty option.

If we do decide to go with an exception 👍 to UOE... unless the NotImplemented thing really is common.